Pull request for Sarina's review

seanlip opened the pull request.

seanlip committed:
Add the actual XBlock.

seanlip committed:
Update README file.

seanlip committed:
Clarify the README further.

nedbat commented on oppia/oppia.py:
```diff
@@ -0,0 +1,132 @@
+    @XBlock.json_handler
+    def on_exploration_loaded(self, data, suffix=''):
+        """Called when an exploration has loaded."""
+        self._log('Exploration %s was loaded.' % self.oppiaid)
```
I'm wondering about these handlers that call _log, and _log doesn't do anything.  Usually, we want analytics events from student activity.  There's a Javascript interface for logging events in XBlocks:  Logger.log, have you seen it?

seanlip replied:
Ah, sorry, I haven't -- thanks for the info. Happy to use it. Do you have a pointer to the documentation, and is this testable using the XBlock workbench?

nedbat replied:
It doesn't look like we have docs on Logger.log, but it is used by XBlocks inside edx-platform.  You don't need to use it, but we should have an event stream somehow.

nedbat replied:
Hmm, here's some light docs on Logger.log:  http://edx.readthedocs.org/projects/edx-developer-guide/en/latest/analytics.html

seanlip replied:
Thanks, this has been fixed.

sarina commented on oppia/oppia.py:
```diff
@@ -0,0 +1,132 @@
+        html = self.resource_string("static/html/oppia.html")
+        frag = Fragment(html.format(self=self))
+        frag.add_javascript_url(
+            "//cdn.jsdelivr.net/oppia/0.0.1/oppia-player.min.js")
```
Who controls this CDN? What code is on it? How much do we trust that this cdn will always be up, and that this will always point at safe code?

seanlip replied:
See http://www.jsdelivr.com/. Note that we also have a script hosted on https://cdnjs.com/.

With regards to these questions, I think that's your call. I'd be happy for the script to be hosted elsewhere as well -- for example, can we do it from edx.org itself? If so, what would be the best way to refer to it?

sarina replied:
@nedbat thoughts?

nedbat replied:
I think we would prefer to have the .min.js be in this repo, and served through our typical deployment mechanisms.  I'm not sure we have a way to do that that wouldn't break on the workbench.

seanlip replied:
Hmm, so, just to check -- what should I do here? I'm not sure I understand what "typical deployment mechanisms" means.

I can certainly bring the .min.js into the repo. Should it go in lib/, oppia/lib, or oppia/static/js? Also, once I've brought that file in, is there anything else I should do?

sarina replied:
I'd like @nedbat to weigh in here; he's out today but I will make sure early next week he gives an answer.

nedbat replied:
The simplest thing is to put the file in oppia/static/js, then access it as you've been getting your other static assets.  That way, the Javascript travels with the Python kit, and will be served by our servers.

seanlip replied:
Hm, that sounds good, but I have a question: is there a way to say "get this file at most once per edX page" (even though the page may contain a couple of Oppia xblocks)? With this proposal I think the JS would be loaded twice, which isn't so nice.

nedbat replied:
It isn't ideal, it's true, but we'd rather have the XBlocks use a uniform mechanism that we can improve later.

seanlip replied:
Fixed; added the file locally as @nedbat suggested.

sarina commented:
1. I'm concerned there's nothing to display in Studio. Even a message indicating that no preview is available would be better than this:

![oppia_studio_name](https://cloud.githubusercontent.com/assets/1985317/9510761/99b8883c-4c3d-11e5-91a7-8df373aa933a.png)

Additionally can you create a better title, maybe "Oppia Block" or "Oppia Unit"? "41de036..." is a terriblly non informative title.
1. In the example block you have me change the height from a default of 500 to 1000. Is this specific to the block, or should the default be 1000?
2. I'm concerned about having to press "Enter" to submit answers. I was confused because I was looking for a "Submit" button and only tried pressing "Enter" as a last resort.

This is made doubly confusing because certain input types - eg the feedback box at the end - actually do have "Submit" buttons. Can you add submit buttons to every input, since they are clearly available?

Additionally, the example block has an error and I got stuck and could no longer proceed. Could you fix this? Screenshot:

![screen shot 2015-08-26 at 12 08 36 pm](https://cloud.githubusercontent.com/assets/1985317/9510774/a5061e02-4c3d-11e5-83b7-f19129913c6b.png)

Finally, about saving and scoring. I'm surprised that your progress on the block doesn't save when you leave or refresh the page. Is that expected use? I would expect that to be frustrating to students, to not be able to come back to the same place in the Oppia unit. You're not passing a score back at all, so is this is just a teaching tool?

cc @explorerleslie - any thoughts?

sarina commented:
Above comment was addressed in email. These blocks are meant to be small, standalone and transient. Think of them like videos, which also don't remember where you left off when you last visited them. (One day we may change this, but that's how it currently is.) A score is not passed back.

As far as the other feedback goes, Sean will look into the Studio display issues, and consider the feedback on "submit" buttons in the Oppia product. As far as the "error', you must enter in one value at a time; I added them all at once. This could be described more clearly in the product.

seanlip commented:
Thank you for the comments! Just wanted to let you know that I'm going to be away for a couple weeks, but will look into the remaining ones when I get back.

sarina commented:
@seanlip great thanks. I think we've just got a few minor things to iron out with the code itself, so I'm confident in that. The accessibility team will do their audit and have the results when you get back. May I ask when that will be, so I can make sure we have full results completed in advance?

seanlip commented:
Hi @sarina, I'll be back on 11 Sep, and am hoping to take a look at this over Sep 12 - 13.

Thank you for the quick responses!

downzer0 commented:
Hi @seanlip. Nice work here so far. My accessibility report is below:
- The `iframe` needs an appropriate `title` attribute. I recommend "Project Euler Problems"
- The "Project Euler Problem 1" image is missing an `alt` attribute
- The "feedback" icon in the top right of each panel doesn't work with a keyboard. Please do the following:
  - Make this control a button, since it's "doing something". Do this for every "feedback" mechanism in the module.
  - Have the button text contain text for screen readers onluy which describes its action. This will hide it visually, but still announce to screen readers the buttons intention. You can do this by applying the CSS class '.sr' to the text wrapper.
  - Place `aria-hidden="true"` on the `span` icon itself.
  - Ensure the control works with a keyboard, and focus is sent to the comment/feedback window when it appears, and is returned to the initial button when the feedback window is closed.
- Furthermore, the "close" button in the feedback control needs to be a button and work with a keyboard.
- The black and white "Oppia bullet" images are missing `alt` attributes. Why not just use normal bullets here? If these aren't valuable to the content (i.e., are just bullets and you'd rather not use normal bullets, please give them a blank `alt` attribute by `alt=""`)
- The green and white "user bullet" images are missing `alt` attributes. Why not just use normal bullets here? If these aren't valuable to the content (i.e., are just bullets and you'd rather not use normal bullets, please give them a blank `alt` attribute by `alt=""`)
- After one of the two choice buttons (solve directly or start smaller) are chosen, focus needs to be sent to the new content that appears. This content should also have a heading indicating what the content is.
- The "title" of the document (in the browser tab, for example) needs to be named something other than a string of numbers and letters. If screen reader users are switching between tabs, what it is now won't make any sense.
- As Sarina mentioned, it's unclear how to actually submit an answer. Utilizing 'Submit' buttons for each problem would s solve this.
- The only indication that a typed response is incorrect is a red line instead of a blue one, but I didn't really notice this at first and became a bit frustrated. I recommend doing some or all of the following:
  - Include a "tip" or "helper" text that indicates the required format in a user friendly fashion. Something humanized and understandable.
  - Since this module appears to check the response as the user types, you'll want to:
    - Provide a second (or even a third) method of feedback that doesn't depend on color alone. Text would be great.
    - Send focus to this additional level of textual feedback (if it's available) when the response is submitted.
- When using the code editor when writing a small program to solve the problem, clicking the "Run" button should send focus to an error message if one exists. Currently, there's no indication that the problem was written correctly or not. Furthermore, the error message should provide useful critique of what happened and suggest how to fix or correct it.
- Please provide a way to "reset" the options. For example, if I begin writing a program but decide I'd rather do a different path, there's currently no way to do this without refreshing the page.

Once you address these issues I'll take another spin. Let me know if you need further explanation or need guidance with them.

seanlip commented:
@sarina -- re the "41de036..." issue, how do I access that value? I don't think I set it.

Also, for modifying the preview, should I add an author_view as described [here](http://edx-developer-guide.readthedocs.org/en/latest/extending_platform/xblocks.html#studio)?

seanlip commented:
@clrux, thank you for the review! Just one question for now:

> The "title" of the document (in the browser tab, for example) needs to be named something other than a string of numbers and letters. If screen reader users are switching between tabs, what it is now won't make any sense.

At the moment, AFAICT, the title of the document is "Project Euler Problem 1 - Oppia". This is what shows up on my browser tab in Chrome. Please could you elaborate on where you see the string of alphanumeric characters?

sarina commented:
@seanlip You need to set a display_name on your xblock. Both Chris and I mentioned the same issue. It's the title on the xblock within the system. http://xblock.readthedocs.org/en/latest/fields.html

sarina commented:
@seanlip Adding an author view would be nice. You could experiment with using this utility package to add the authoring mixin: https://github.com/edx/xblock-utils

sarina commented:
Also, have we discussed analytics + documentation?

sarina commented:
@seanlip Hi, how's it going? Let me know please when you'd like us to take another look at this block.

Note that I'm OOO 9/24 - 9/29, but would like to help in what ways I can around these dates. Thanks!

seanlip commented:
Hi @sarina, thanks for checking in! This xblock isn't ready to take another look at yet, but it will be soon.

There are two categories of changes I need to make. One is to the XBlock itself, and one is to the Oppia student UI (many of the accessibility fixes fall in the latter category). I've finished most of the student UI fixes, and we are planning to do a release later this week with the changes. On the other hand, I plan to tackle the XBlock-related changes in one batch, and the main thing I would need is a response about what to do with the .min.js script.

Also, thanks for the OOO notification! I don't think the XBlock changes will be in before you leave, but I'll try to get it ready by the time you return.

seanlip committed:
Add display_name field and author_view method; remove width/height fields; add event tracking; include a local copy of the embedding script with an appropriate title for the iframe.

seanlip committed:
Update README.md

seanlip commented:
Thank you @sarina, @nedbat and @clrux for the comments. I have addressed them in the latest commits for this XBlock, and some of the accessibility comments are also addressed in the new release of Oppia that we pushed out yesterday.

In particular, I have added a `display_name` and a simple `author_view` (for some reason, the default student_view preview does not work in Studio), which should make things clearer for course creators. I've also removed the 'width' and 'height' fields, in order to keep the embedding experience consistent throughout.

With regards to the accessibility comments, we have:
- Added a title to the iframe
- Removed the image at the top
- Made the changes @clrux suggested for the feedback icon in the top right of each card
- Added empty alt attributes to the bullet images
- Redesigned the multiple choice, numeric input and text input interactions

In addition, I should mention that we don't currently have functionality for 'resetting' an answer. The current intention is for explorations to play out a bit like 'choose your own adventure' games, which don't usually have options for backtracking within the story. We might revisit this at a later stage, but I'm afraid it probably won't happen anytime soon.

Again, thank you for all the comments!

sarina commented on oppia/static/html/oppia_preview.html:
```diff
@@ -0,0 +1,10 @@
+<div style="opacity: 0.7;">
+  <p>
+    This XBlock does not have an inline Studio preview.
```
Is this supposed to show up in Studio? I don't see it:

![screen shot 2015-09-30 at 1 47 20 pm](https://cloud.githubusercontent.com/assets/1985317/10201386/cbc77908-6779-11e5-963d-6426b573ef5f.png)

seanlip replied:
That's odd. It does for me.

What happens if you refresh the page?

sarina replied:
Yeah, it shows up on refresh. Hm.

sarina commented on oppia/oppia.py:
```diff
@@ -0,0 +1,148 @@
+    _EVENT_NAME_STATE_TRANSITION = 'oppia.exploration.state.changed'
+
+    # The display name of the component. Note that this is not editable in
+    # Studio.
```
Why can't I edit this in Studio?

Instructors will want to edit this so I'd like to discuss your reasoning for not allowing edits here.

seanlip replied:
I thought this was just the internal title of the block in Studio ("Oppia Exploration" in your screenshot above)? I looked at some other XBlocks, and didn't see this field in their studio edit_html file.

The comment is here mainly to distinguish this field from the other ones.

sarina replied:
This should be editable in Studio. This is the name that shows up on the unit in the LMS tooltip:

![screen shot 2015-10-05 at 4 49 18 pm](https://cloud.githubusercontent.com/assets/1985317/10293160/12f58814-6b81-11e5-8c31-d039d25d3001.png)

If a course has a sequence with more than one Oppia exploration, they will want to edit the names of the explorations to make them distinct. Even if they don't have more than one, they may want to call this eg "Quadratic Formula Oppia Exploration" to make it clear what you're exploring.

sarina replied:
This field is editable in all XBlocks; try adding any problem type and you'll see that while there's always a default title, there's also the option to edit it:

![screen shot 2015-10-05 at 4 51 19 pm](https://cloud.githubusercontent.com/assets/1985317/10293223/5b891172-6b81-11e5-9a6c-468f688264f6.png)

seanlip replied:
That's odd, I didn't actually do anything special to try and hide the field. Are you saying I should add an additional custom field (similar to the ones that currently exist for src and id) in the editor view? If so, yep, I can do that.

sarina replied:
Ugh that's frustrating. No, let me go see what other block types do. Sorry I don't know offhand.

sarina commented:
This is looking really good! I have a few minor concerns, some noted inline above, and a few more:

I get stuck in the Python editor. The "RUN" button is greyed out. Is this expected? if so I'd recommend just removing the code part, as it's not a good user experience.

![screen shot 2015-09-30 at 1 51 59 pm](https://cloud.githubusercontent.com/assets/1985317/10201535/902c1538-677a-11e5-8cb7-bfa2881a9b7a.png)

Secondly, I will work with Chris to get another a11y review going; in the meantime I want to make sure helper text, such as "Type a number" here, is accessible:

![screen shot 2015-09-30 at 1 53 33 pm](https://cloud.githubusercontent.com/assets/1985317/10201566/b60310fe-677a-11e5-96ef-e4bd44a336b4.png)

downzer0 commented:
@seanlip Would you mind spinning up a sandbox for me to preview this? Or, if not a sandbox, some way (even instructions) for me to preview?

sarina commented:
@clrux I already did this, information is on the JIRA ticket. Most open source contributors cannot spin up a sandbox.

seanlip commented:
Thanks @sarina!

The Python issue is a bug that appears to manifest only on some browsers (notably, v45 of Chrome). We're aware of this and are looking into fixing it. This should, however, not affect the upcoming course, which doesn't use the code interaction. (For that course, the exploration id is **kJzlCWmP8utj**; this might be a better one to use instead for accessibility testing.)

Also, I would expect the placeholder text to be accessible, but perhaps @clrux could confirm.

downzer0 commented:
@seanlip Nice work. There are a few more finicky things I've found.
- The feedback button and icon seem to be separate. There's an off-screen button that's actionable, then an on-screen "comment" icon which just says "button". There's no indication of what the "button" does.
  <img width="186" alt="screen shot 2015-10-02 at 2 02 32 pm" src="https://cloud.githubusercontent.com/assets/2112024/10254579/af5be840-6912-11e5-9e00-94ad5f120bd8.png">
- The navigation arrows are icon fonts without any context. We have an `.sr-only` class which will let you put text next to icons which is hidden visually, but announced for screen readers.
  <img width="194" alt="screen shot 2015-10-02 at 2 03 02 pm" src="https://cloud.githubusercontent.com/assets/2112024/10254603/d78ded36-6912-11e5-98a5-6719b70a6e0f.png">

```
<span class="icon" aria-hidden="true"></span>
<span class="sr-only">Back</span>
```
- The "I want to write a simple program..." text isn't clickable after clicking the first time and navigating backwards.
  <img width="574" alt="screen shot 2015-10-02 at 2 37 13 pm" src="https://cloud.githubusercontent.com/assets/2112024/10254651/195f6df2-6913-11e5-82d0-8891877c0689.png">
- The text "(enter text in the code editor)" text is a button, but it isn't indicated as such. There's no clear direction what to do. Could you make this a button?

I think that's all on this pass. Making good progress!

seanlip commented:
Thanks @clrux! I have one question and one clarification. (I've also addressed the other two points, and they should be fixed in the next Oppia release.)

> The feedback button and icon seem to be separate. There's an off-screen button that's actionable, then an on-screen "comment" icon which just says "button". There's no indication of what the "button" does.

I've looked at the HTML of the page, and I can't see an off-screen button. Could you please clarify which button this is? I also double-checked and we have set an [sr-only](https://github.com/oppia/oppia/blob/master/extensions/skins/conversation_v1/player.html#L42) attribute as you advised previously, so I'm not sure where the 'button' text is coming from...

> The "I want to write a simple program..." text isn't clickable after clicking the first time and navigating backwards. 

I'm afraid this isn't going to be fixed: we don't currently support 'resetting' an answer. As mentioned in a previous comment, the current intention is for explorations to play out a bit like 'choose your own adventure' games or one-on-one tutor conversations, which don't usually have options for backtracking within the story unless the author of the exploration explicitly enables this. We might revisit this at a later stage, but it probably won't happen anytime soon.

downzer0 commented:
@seanlip Great question. See the screenshot below and then I'll explain what's going on.

<img width="714" alt="screen shot 2015-10-05 at 8 27 04 am" src="https://cloud.githubusercontent.com/assets/2112024/10280397/ea7a2912-6b3a-11e5-803a-019d1ac10a33.png">

We have a `button` but it's only "text" is an icon that's hidden from screen readers. So, without any text here, all screenreaders know to announce is the button itself without a description. Try this:

```
<button class="conversation-skin-tutor-card-feedback-toggle" popover-placement="left" popover-template="popover/feedback" popover-trigger="click" tooltip="Feedback" tooltip-placement="bottom" state-name="Intro">
    <span class="sr-only">Leave feedback on this card</span>
    <span class="glyphicon glyphicon-comment" aria-hidden="true"></span>
</button>
```

This will provide content/a label for the button (Leave feedback...) while maintaining the current look. Note that I've moved the `span` that's above the `button` into the `button` as its content.

downzer0 commented:
@seanlip One more thing, we use Font Awesome in our platform, so you may want to change the icon `span` to:

```
<span class="icon fa fa-comment" aria-hidden="true"></span>
```

If you're including an additional font icon library, I may suggest seeing if there are any equivalent in Font Awesome (4.3) to simplify and reduce dependencies.

sarina commented on oppia/oppia.py:
```diff
@@ -0,0 +1,148 @@
+    display_name = String(
+        help="Display name of the component",
+        default="Oppia Exploration",
+        scope=Scope.settings)
```
So, I think the issue here is that you're specifying scope.settings.

This XBlock has an editable display name, and defines it like so: `display_name = String(default='Poll')` 

![screen shot 2015-10-05 at 6 29 43 pm](https://cloud.githubusercontent.com/assets/1985317/10295524/30964594-6b8f-11e5-9f01-3ac656a8a6b7.png)

https://github.com/open-craft/xblock-poll/blob/master/poll/poll.py#L275

So either remove the scope declarator, or change to scope.content.

sarina commented:
@seanlip Ball is in your court right now but I think we're really close. From my perspective the remaining TODOs are:
- Make sure the display name is editable
- Respond to the few outstanding a11y concerns (let us know what you can do; @clrux suggested some edx-platform specific CSS tweaks, but not sure that Oppia is going to make use of those. However, you may be able to check out the edx-platform CSS and get ideas of how to make Oppia more accessible)

seanlip committed:
Add a field in Studio for the component's display name.

seanlip committed:
Merge branch 'master' of https://github.com/oppia/xblock

seanlip commented:
Thanks, @clrux! I have made the fix to the `sr-only` tag. It will go out in the next Oppia release, which is likely to be on or around Monday 12th Oct.

@sarina, I looked at the XBlock you link to and it looks like they actually do add an additional custom field in the editor view after all, so I just went ahead and did that. The display name is now editable.

I think this should take care of everything.

sarina commented:
Hi @seanlip and thanks! I will test out the XBlock changes today.

@clrux does this take care of your a11y concerns?

downzer0 commented:
The only outstanding thing is to move the `<span class="sr-only">Leave feedback on this card</span>` inside of the feedback `button`. As it is now, the `button` only announces as "button" because the `sr` text is outside of the element.

sarina commented:
@clrux Yes, Sean indicated that this has been fixed in Oppia (the block is a view of the Oppia tool, so it has to be fixed within that tool). I was clarifying that beyond that, there's no other fix we're waiting on

downzer0 commented:
Perfect.

sarina commented:
@seanlip Arugh. I'm all over the place. I don't think we talked about analytics eventing. I think the block probably ought to emit an event but would welcome your thoughts. Our guidelines are here: https://openedx.atlassian.net/wiki/display/OPEN/XBlock+Analytics+Guidelines

I'm happy to take a deeper dive on this chat via email. I sent you an email this morning regarding documentation of this block as well.

seanlip commented:
@sarina I think the block already emits events? See the oppia.py file...

sarina commented on oppia/oppia.py:
```diff
@@ -0,0 +1,143 @@
+        """
+        Logger for load, state transition and completion events.
+        """
+        tracker.emit(event_name, payload)
```
Our analytics team recommends that you use `self.runtime.publish`, not `tracker.emit` directly.

From them: "so they are basically just bypassing the runtime and calling "tracker.emit" directly, which will work, but only in the LMS.
if you tried to run that XBlock in another xblock runtime it would fail, where as "publish" allows the runtime to define the behavior.
ideally we would expose a more comprehensive instrumentation "service" to the xblock, but we haven't gotten around to building such a thing
And more importantly, going through our runtime allows us to set some context on the tracking events being emitted."

sarina replied:
Here's an example: https://github.com/edx/edx-platform/blob/ceb5655b2d3520129ccd363c1caef6b8c3676ad1/common/lib/xmodule/xmodule/library_content_module.py#L147

seanlip replied:
Done. Thanks!

seanlip committed:
Change tracker.emit to self.runtime.publish, per Sarina's advice.

sarina commented:
@seanlip I'm done with this PR :joy: 

You can go ahead and close it. Any updates that you have to the block, please make in a PR (rather than committing straight to master) so we can do a quick review of it.

What's remaining to do for this I've sent in an email; our next production release is 10/19 and I will make sure the current version of the block is installed. If you end up making any more commits to this block keep me in the loop so I can be sure to update the installed version.

seanlip commented:
SGTM. Thanks @sarina!

Just a thought -- given what you mentioned in the second paragraph, it might be a good idea to watch the repo, too.

sarina commented:
Sure thing. We install the XBlock by hash, though, so we'll install the hash `cd5479e` at this point. If you make changes, you'll have to let us know so we can upgrade the hash (for security reasons, etc etc)

seanlip commented:
Ah, yes, makes sense. Thanks!
